File: ReactAndroid/src/main/java/com/facebook/react/uimanager/ShadowNodeRegistry.java
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package com.facebook.react.uimanager;

import android.util.SparseArray;
import android.util.SparseBooleanArray;
import com.facebook.react.common.SingleThreadAsserter;

/**
 * Simple container class to keep track of {@link ReactShadowNode}s associated with a particular
 * UIManagerModule instance.
 */
public class ShadowNodeRegistry {

  private final SparseArray<ReactShadowNode> mTagsToCSSNodes;
  private final SparseBooleanArray mRootTags;
  private final SingleThreadAsserter mThreadAsserter;

  public ShadowNodeRegistry() {
    mTagsToCSSNodes = new SparseArray<>();
    mRootTags = new SparseBooleanArray();
    mThreadAsserter = new SingleThreadAsserter();
  }

  public void addRootNode(ReactShadowNode node) {
    // TODO(6242243): This should be asserted... but UIManagerModule is
    // thread-unsafe and calls this on the wrong thread.
    //mThreadAsserter.assertNow();
    int tag = node.getReactTag();
    mTagsToCSSNodes.put(tag, node);
    mRootTags.put(tag, true);
  }

  public void removeRootNode(int tag) {
    mThreadAsserter.assertNow();
    if (!mRootTags.get(tag)) {
      throw new IllegalViewOperationException(
          "View with tag " + tag + " is not registered as a root view");
    }

    mTagsToCSSNodes.remove(tag);
    mRootTags.delete(tag);
  }

  public void addNode(ReactShadowNode node) {
    mThreadAsserter.assertNow();
    mTagsToCSSNodes.put(node.getReactTag(), node);
  }

  public void removeNode(int tag) {
    mThreadAsserter.assertNow();
    if (mRootTags.get(tag)) {
      throw new IllegalViewOperationException(
          "Trying to remove root node " + tag + " without using removeRootNode!");
    }
    mTagsToCSSNodes.remove(tag);
  }

  public ReactShadowNode getNode(int tag) {
    mThreadAsserter.assertNow();
    return mTagsToCSSNodes.get(tag);
  }

  public boolean isRootNode(int tag) {
    mThreadAsserter.assertNow();
    return mRootTags.get(tag);
  }

  public int getRootNodeCount() {
    mThreadAsserter.assertNow();
    return mRootTags.size();
  }

  public int getRootTag(int index) {
    mThreadAsserter.assertNow();
    return mRootTags.keyAt(index);
  }
}

File: ReactWindows/ReactNative/UIManager/ShadowNodeRegistry.cs
using System;
using System.Collections.Generic;

namespace ReactNative.UIManager
{
    /// <summary>
    /// Simple container class to keep track of <see cref="ReactShadowNode"/>s
    /// associated with a particular <see cref="UIManagerModule"/> instance.
    /// </summary>
    public class ShadowNodeRegistry
    {
        private readonly IDictionary<int, ReactShadowNode> _tagsToCssNodes =
            new Dictionary<int, ReactShadowNode>();

        private readonly IDictionary<int, bool> _rootTags =
            new Dictionary<int, bool>();

        /// <summary>
        /// The collection of root node tags.
        /// </summary>
        public ICollection<int> RootNodeTags
        {
            get
            {
                return _rootTags.Keys;
            }
        }  

        /// <summary>
        /// Add a root shadow node.
        /// </summary>
        /// <param name="node">The node.</param>
        public void AddRootNode(ReactShadowNode node)
        {
           if (node == null)
                throw new ArgumentNullException(nameof(node));

            var tag = node.ReactTag;
            _tagsToCssNodes[tag] = node;
            _rootTags[tag] = true;
        }

        /// <summary>
        /// Remove a root shadow node.
        /// </summary>
        /// <param name="tag">The tag of the node to remove.</param>
        public void RemoveRootNode(int tag)
        {
            if (!_rootTags.ContainsKey(tag))
            {
                throw new KeyNotFoundException(
                    $"View with tag '{tag}' is not registered as a root view.");
            }

            _tagsToCssNodes.Remove(tag);
            _rootTags.Remove(tag);
        }

        /// <summary>
        /// Add a react shadow node.
        /// </summary>
        /// <param name="node">The node to add.</param>
        public void AddNode(ReactShadowNode node)
        {
            if (node == null)
                throw new ArgumentNullException(nameof(node));

            _tagsToCssNodes[node.ReactTag] = node;
        }

        /// <summary>
        /// Remove a react shadow node.
        /// </summary>
        /// <param name="tag">The tag of the node to remove.</param>
        public void RemoveNode(int tag)
        {
            var isRoot = default(bool);
            if (_rootTags.TryGetValue(tag, out isRoot) && isRoot)
            {
                throw new KeyNotFoundException(
                    $"Trying to remove root node '{tag}' without using RemoveRootNode.");
            }

            _tagsToCssNodes.Remove(tag);
        }

        /// <summary>
        /// Retrieve a react shadow node.
        /// </summary>
        /// <param name="tag">The tag of the node to retrieve.</param>
        /// <returns>The react shadow node.</returns>
        public ReactShadowNode GetNode(int tag)
        {
            var result = default(ReactShadowNode);
            if (_tagsToCssNodes.TryGetValue(tag, out result))
            {
                return result;
            }

            throw new KeyNotFoundException(
                $"Shadow node for tag '{tag}' does not exist.");
        }

        /// <summary>
        /// Checks if a node with the given tag is a root node.
        /// </summary>
        /// <param name="tag">The tag.</param>
        /// <returns>
        /// <b>true</b> if the node with the given tag is a root node,
        /// <b>false</b> otherwise.
        /// </returns>
        public bool IsRootNode(int tag)
        {
            return _rootTags.ContainsKey(tag);
        }
    }
}
